### PR TITLE
Fixed error when passing shell reserved characters to setup databases 

### LINKED ISSUE #4925: cloudstack-setup-databases fails on complex database key
##### ISSUE TYPE
 * Bug Report

##### COMPONENT NAME
~~~
Setup
~~~

##### CLOUDSTACK VERSION
~~~
4.15.0
~~~

##### CONFIGURATION
not relevant I think

##### OS / ENVIRONMENT
not relevant I think

##### SUMMARY
I run 
cloudstack-setup-databases cloud:<dbpassword>@localhost --deploy-as=root:<password> -e <encryption_type> -m <management_server_key> -k <database_key> -i <management_server_ip>
with e.g. the following (see database key)
cloudstack-setup-databases cloud:<dbpassword>@localhost --deploy-as=root:<password> -e <encryption_type> -m <management_server_key> -k "Some&Crazy&Password" -i <management_server_ip>
it splits the key and try to run it as commands (see screenshot)
Bildschirmfoto 2021-04-15 um 15 27 36

##### STEPS TO REPRODUCE

1. Follow install instructions on https://docs.cloudstack.apache.org/en/4.15.0.0/installguide/management-server/index.html
2. in step where you have to run cloudstack-setup-databases, do something like the following:

~~~
cloudstack-setup-databases cloud:"password"@localhost --deploy-as=root:"password" -m "password" -k "Some&Crazy&Password"
~~~

##### EXPECTED RESULTS
~~~
Script should setup the given key.
~~~

##### ACTUAL RESULTS
Error message when processing encryption.

~~~
see screenshot above
~~~

### PR DESCRIPTION
**Pull request by Spaceman1984**
### Description

This PR will allow reserved Linux shell characters to be passed to the setup databases script for the database secret key.
Fixes #4925

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [x] Minor
- [ ] Trivial

### How Has This Been Tested?

This can be tested by executing the script something like this

`cloudstack-setup-databases cloud:"password"@localhost --deploy-as=root:"password" -m "password" -k "Some&Crazy&Password"`

### CODE REVIEW

**Comment by Spaceman1984** [comment]
@blueorangutan package

**Comment by Spaceman1984** [comment]
@blueorangutan test

File: setup/bindir/cloud-setup-databases.in [hunk 1]
```diff
@@ -26,6 +26,7 @@ import string
 from optparse import OptionParser
 import subprocess
 import shutil
+import shlex
```
Comment (rhtyd) [inline setup/bindir/cloud-setup-databases.in:29]: @Spaceman1984 what installs `shlex`? It does not look like part of standard py3 library.

Comment (Spaceman1984) [reply setup/bindir/cloud-setup-databases.in:29]: This is standard python

**Comment by rhtyd** [comment]
@Spaceman1984 pl change base branch of the PR to 4.15 and rebase your PR against 4.15

File: setup/bindir/cloud-setup-databases.in [hunk 2]
```diff
@@ -443,7 +444,7 @@ for example:
 
             self.encryptiontype = self.options.encryptiontype
             self.mgmtsecretkey = self.options.mgmtsecretkey
-            self.dbsecretkey = self.options.dbsecretkey
+            self.dbsecretkey = shlex.quote(self.options.dbsecretkey)
```
Comment (rhtyd) [inline setup/bindir/cloud-setup-databases.in:447]: @Spaceman1984 this adds a quote to the string, why do we want to add that? Are we passing it as a parameter to something on shell?

Comment (Spaceman1984) [reply setup/bindir/cloud-setup-databases.in:447]: Yes, the ampersand breaks a call to the shell. Refer back to the error on #4925

Comment (rhtyd) [reply setup/bindir/cloud-setup-databases.in:447]: Pl see the usage in `encrypt()` and add quotes there.

**Comment by Spaceman1984** [comment]
@blueorangutan package

**Comment by Spaceman1984** [comment]
@blueorangutan test

**Review by Pearl1594 — APPROVED** [review]
Tested - LGTM
#### Prior Fix:
```
[root@trl-176-k-M7-pearl-dsilva-mgmt1 ~]# cloudstack-setup-databases cloud:"P@ssword123"@localhost --deploy-as=root:"P@ssword123" -m "P@ssword123" -k "Some&Crazy&Password"
Mysql user name:cloud                                                           [ OK ]
Mysql user password:******                                                      [ OK ]
Mysql server ip:localhost                                                       [ OK ]
Mysql server port:3306                                                          [ OK ]
Mysql root user name:root                                                       [ OK ]
Mysql root user password:******                                                 [ OK ]
Checking Cloud database files ...                                               [ OK ]
Checking local machine hostname ...                                             [ OK ]
Checking SELinux setup ...                                                      [ OK ]
Detected local IP address as 10.0.33.183, will use as cluster management server node IP[ OK ]
Preparing /etc/cloudstack/management/db.properties                              [ OK ]
Applying /usr/share/cloudstack-management/setup/create-database.sql             [ OK ]
Applying /usr/share/cloudstack-management/setup/create-schema.sql               [ OK ]
Applying /usr/share/cloudstack-management/setup/create-database-premium.sql     [ OK ]
Applying /usr/share/cloudstack-management/setup/create-schema-premium.sql       [ OK ]
Applying /usr/share/cloudstack-management/setup/server-setup.sql                [ OK ]
Applying /usr/share/cloudstack-management/setup/templates.sql                   [ OK ]
Processing encryption ...                                                       Traceback (most recent call last):
  File "/usr/bin/cloudstack-setup-databases", line 604, in <module>
    o.run()
  File "/usr/bin/cloudstack-setup-databases", line 593, in run
    self.processEncryptionStuff()
  File "/usr/bin/cloudstack-setup-databases", line 418, in processEncryptionStuff
    encryptDBSecretKey()
  File "/usr/bin/cloudstack-setup-databases", line 402, in encryptDBSecretKey
    self.putDbProperty('db.cloud.encrypt.secret', formatEncryptResult(encrypt(self.dbsecretkey)))
  File "/usr/bin/cloudstack-setup-databases", line 392, in encrypt
    return str(runCmd(cmd)).strip('\r\n')
  File "/usr/bin/cloudstack-setup-databases", line 51, in runCmd
    raise Exception(stderr)
Exception: b'/bin/sh: Password: command not found\n/bin/sh: Crazy: command not found\n\nUSAGE: encrypt.sh [ARGUMENTS]\n\n  * Arguments must apply to format:\n\n      "arg1=value1 arg2=value2 arg3=value3 ..."\n\n  * Required arguments:\n\n      input\n      password\n\n  * Optional arguments:\n\n      verbose\n      algorithm\n      keyObtentionIterations\n      saltGeneratorClassName\n      providerName\n      providerClassName\n      stringOutputType\n      ivGeneratorClassName\n'

```

#### Post Fix:
```
[root@trl-176-k-M7-pearl-dsilva-mgmt1 ~]# cloudstack-setup-databases cloud:"P@ssword123"@localhost --deploy-as=root:"P@ssword123" -m "P@ssword123" -k "Some&Crazy&Password"
Mysql user name:cloud                                                           [ OK ]
Mysql user password:******                                                      [ OK ]
Mysql server ip:localhost                                                       [ OK ]
Mysql server port:3306                                                          [ OK ]
Mysql root user name:root                                                       [ OK ]
Mysql root user password:******                                                 [ OK ]
Checking Cloud database files ...                                               [ OK ]
Checking local machine hostname ...                                             [ OK ]
Checking SELinux setup ...                                                      [ OK ]
Detected local IP address as 10.0.33.183, will use as cluster management server node IP[ OK ]
Preparing /etc/cloudstack/management/db.properties                              [ OK ]
Applying /usr/share/cloudstack-management/setup/create-database.sql             [ OK ]
Applying /usr/share/cloudstack-management/setup/create-schema.sql               [ OK ]
Applying /usr/share/cloudstack-management/setup/create-database-premium.sql     [ OK ]
Applying /usr/share/cloudstack-management/setup/create-schema-premium.sql       [ OK ]
Applying /usr/share/cloudstack-management/setup/server-setup.sql                [ OK ]
Applying /usr/share/cloudstack-management/setup/templates.sql                   [ OK ]
Processing encryption ...                                                       [ OK ]
Finalizing setup ...                                                            [ OK ]

CloudStack has successfully initialized database, you can check your database configuration in /etc/cloudstack/management/db.properties

```

**Review by rhtyd — APPROVED** [review]

**Pull request merged**